### PR TITLE
feat(firestore): cache snapshot connections with underlying native listener

### PR DESCRIPTION
**Pull request by rrousselGit**
## Description

Currently, calling `.snapshots()` twice on the same reference leads to listening the same reference twice, which is a billable operation.

This is a common issue when using reference like so:

```dart
StreamBuilder(
  stream: collection.where(...).snapshots(),
  ...
)
```

where rebuilding the StreamBuilder with the exact same query leads to a "removeListener" followed by "addListener" on the query, which is expensive.

This PR caches the bridge with the native platform so that when rebuilding a StreamBuilder like previously mentioned, this will instead reuse the existing connection instead of opening a new one and closing the previous one.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cloud_firestore/cloud_firestore/lib/src/snapshot_cache.dart [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    // StreamBuilder(stream: reference.snapshots(), ...)
+    //
+    // In this scenarion, StreamBuilder would cancel the previous stream
+    // subscription then listen to the new stream.
+    // This would call the previous controller's onCancel **before** the new
+    // controller's onListen. Which would clear the native bridge cache
+    // when the expected behaviour is to reuse the connection.
+    return Future(() {
+      connectionSub?.cancel();
+      // if (!controllerWithLastEvent.isClosed) {
```
Comment (Ehesp) [inline packages/cloud_firestore/cloud_firestore/lib/src/snapshot_cache.dart:106]: Just double checking, is this if statement supposed to be commented out?

Comment (rrousselGit) [reply packages/cloud_firestore/cloud_firestore/lib/src/snapshot_cache.dart:106]: Ah yes, it's just dead code

File: packages/cloud_firestore/cloud_firestore/lib/src/snapshot_cache.dart [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+          },
+        );
+      };
+
+      nativeController.onCancel = () {
+        sub.cancel();
+        // When all listeners on the native bridge are removed, remove the
+        // entry from cache to avoid memory leaks
+        snapshotCache.remove(key);
+        // nativeController.close();
```
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore/lib/src/snapshot_cache.dart:65]:
```suggestion
        snapshotCache.remove(key);
```

File: packages/cloud_firestore/cloud_firestore/lib/src/snapshot_cache.dart [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    // In this scenarion, StreamBuilder would cancel the previous stream
+    // subscription then listen to the new stream.
+    // This would call the previous controller's onCancel **before** the new
+    // controller's onListen. Which would clear the native bridge cache
+    // when the expected behaviour is to reuse the connection.
+    return Future(() {
+      connectionSub?.cancel();
+      // if (!controllerWithLastEvent.isClosed) {
+      //   controllerWithLastEvent.close();
+      // }
```
Comment (rrousselGit) [inline packages/cloud_firestore/cloud_firestore/lib/src/snapshot_cache.dart:108]:
```suggestion
      connectionSub?.cancel();
```

**Comment by lesnitsky** [comment]
does it make sense to add tests which verify an actual call count of the method channel?

**Review by russellwheatley — APPROVED** [review]
LGTM.

**Comment by rrousselGit** [comment]
> does it make sense to add tests which verify an actual call count of the method channel?

That'd probably be redundant with the tests for `getCachedSnapshot` – which verify that the stream is cached already.

**Review by lesnitsky — APPROVED** [review]

**Review by Salakar — APPROVED** [review]

**Comment by tcf909** [comment]
I do want to raise an issue this commit created.

You've changed the behavior for listeners of this now cached connection.

The first listener will receive a full list of docChanges (with every record being represented as added).

the N+1 listener will only see the most recent updated records in docChanges initially and instead will need to rely on just docs to get a full list of the documents in the snapshot.

Certainly a breaking change.  And goes against the documented behavior (that the initial stream listener event will contain all documents in the docChanges as added).

**Comment by Salakar** [comment]
cc @rrousselGit

**Comment by rrousselGit** [comment]
That's a good point.
I'm not too familiar with `docChanges`. Are we expecting the initial call to simply contains all the "docs" with nothing else?

In particular I'm wondering if we can get this behaviour back without reverting this change.

**Comment by larssn** [comment]
I'm not sure they charge you for re-reads. [The official pricing docs](https://cloud.google.com/firestore/pricing) say this:

```
Listening to query results
Firestore allows you to listen to the results of a query and get realtime updates when the query results change.

When you listen to the results of a query, you are charged for a read each time a document in the result set is added or updated. You are also charged for a read when a document is removed from the result set because the document has changed. (In contrast, when a document is deleted, you are not charged for a read.)

Also, if the listener is disconnected for more than 30 minutes (for example, if the user goes offline), you will be charged for reads as if you had issued a brand-new query.
```

The last paragraph seem to indicate that the snapshot listener has to be disconnected for more than 30min to be charged again.

**Pull request merged**